_posts/2016-12-27-angular-2-advance-testing-with-custom-matchers.md

```diff
@@ -0,0 +1,650 @@
+  });
+
+});
+
+/*
+ * Shell component with property 'direction' that will be used
+ * with our tests
+ */
+@Component({
+  selector: 'test-layout',
+  template: `<span>PlaceHolder HTML to be Replaced</span>`
+})
+export class TestLayoutComponent implements OnInit {
+  direction = "column";
+  constructor() {   }
+  ngOnInit() {   }
+}
+
+/*
+ * Custom Helper function to quickly create a `fixture` instance based on
+ * the 'TestLayoutComponent' class
+ */
+function createTestComponent(template: string): ComponentFixture<TestLayoutComponent>
+{
+  return TestBed
+    .overrideComponent(TestLayoutComponent, {set: {template: template}} )
+    .createComponent(TestLayoutComponent);
+}
+{% endraw %}
+{% endhighlight %}
+
+I now have everything I need to write a Directive test quickly.
+
+{% highlight js %}
+{% raw %}
+it('should compile with custom directives', () => {
+  let fixture = createTestComponent(`<div fxLayout></div>`);
+  expect( fixture ).toBeDefined();
+});
+{% endraw %}
+{% endhighlight %}
+
+Wow! That is pretty easy... and the component has been constructed and prepared with the same 
+processes and DI that your real world application would use.
+
+### 2) Using DRY Helper Methods
+
+Let's first write out test using the traditional long-form... one without custom matchers and the 
+more advanced helper methods.
+
+Since the **fxLayout** directive will add custom flexbox CSS to the host element, our test logic 
+will confirm that the initial CSS is correctly assigned.
+
+###### Test Directive Logic: Long-Form
+
+The traditional approach would probably implement something like this:
+
+![pic4](https://cloud.githubusercontent.com/assets/210413/21525027/fd15cf06-ccdf-11e6-8da2-3ac91d5b9d08.png)
+
+
+
+Above I defined a custom template with bindings to the component property `direction`. Then I use 
+deeply nested references to get access to the native element... and then I test each style individually.
+
+To exacerbate this issue, imagine that our test module has more than 20 individual `it(...)` tests! 
+That is a lot of duplicate code. And there is certainly nothing DRY ("do not repeat yourself") about 
+this code! 
+
+###### Test Directive Logic: Short-Form
+
+Here is the DRY version that I want:
+
+![pic51](https://cloud.githubusercontent.com/assets/210413/21514729/2656a662-cc8e-11e6-8b7a-0594ea3cf120.png)
+
+All the complexities of forcing change detection, accessing the native element, and confirming  
+1...n DOM CSS styles is now easily encapsulated in a Helper function and a Custom Matcher 
+(respectively). 
+
+Instead of using the standard `expect(...)` method, I define a custom *expect* Helper function 
+`expectNativeEl( <fixture> )`:
+
+{% highlight js %}
+{% raw %}
+/**
+ * Note: this helper only accesses the 1st-level child within a component
+ *       A different helper method is need to access deep-level DOM nodes
+ */
+export function expectNativeEl(fixture: ComponentFixture<any>): any {
+  fixture.detectChanges();
```

Comment:
It might be worth mentioned that, since `detechtChanges()` is inside the `expectNativeEL()` function, every time we run `expectNativeEL()`, change detection is performed, even though it's not needed.

E.g. having a test like this:

```
it('should..., () => {
  let fixture = createTestComponent(`<div [fxLayout]="direction"></div>`);

  expectNativeEL(fixture).toHaveCssStyle(...) // Change detection performed
  expectNativeEL(fixture).toHaveSomethingElse(...) // Change detection performed again
  expectNativeEL(fixture).toHaveAnotherThing(...) // Hoppla, again! 3 times now, even though only once was needed
})
```

This problem of course we won't have when we move the detectChanges() call into the spec itself. That makes it a bit more transparent, plus we have more control over when CD is performed.

The more I think about it the more I'm unsure if it's really a good idea to put the CD call inside the helper function.